daisy/workflow/sources.go

```diff
@@ -0,0 +1,114 @@
+package workflow
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+)
+
+func (w *Workflow) recursiveGCS(bkt, prefix, dst string) error {
+	it := w.StorageClient.Bucket(bkt).Objects(w.Ctx, &storage.Query{Prefix: prefix})
+	for objAttr, err := it.Next(); err != iterator.Done; objAttr, err = it.Next() {
+		if err != nil {
+			return err
+		}
+		if objAttr.Size == 0 {
+			continue
+		}
+		srcPath := w.StorageClient.Bucket(bkt).Object(objAttr.Name)
+		o := path.Join(w.sourcesPath, dst, strings.TrimPrefix(objAttr.Name, prefix))
+		dstPath := w.StorageClient.Bucket(w.bucket).Object(o)
+		if _, err := dstPath.CopierFrom(srcPath).Run(w.Ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *Workflow) uploadFile(src, obj string) error {
+	dstPath := w.StorageClient.Bucket(w.bucket).Object(path.Join(w.sourcesPath, obj))
+	gcs := dstPath.NewWriter(w.Ctx)
+	f, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(gcs, f); err != nil {
+		return err
+	}
+	return gcs.Close()
+}
+
+func (w *Workflow) uploadSources() error {
+	for dst, origPath := range w.Sources {
+		// GCS to GCS.
+		if bkt, objPath, err := splitGCSPath(origPath); err == nil {
+			if objPath == "" || strings.HasSuffix(objPath, "/") {
```

Comment:
What do you think about forcing people to specify GCS "folders" by making the trailing "/" required? (instead of doing your ErrObjectNotExist check below)